server/delay_server/src/irodsDelayServer.cpp

```diff
@@ -722,7 +736,19 @@ int main(int argc, char** argv)
     irods::delay_queue queue{queue_size_in_bytes};
 
     try {
-        while (!delay_server_terminated) {
+        while (true) {
+            if (delay_server_terminated) {
+                break;
+            }
+
+            const auto state = irods::server_state::get_state();
+
+            if (state == irods::server_state::server_state::stopped ||
+                state == irods::server_state::server_state::exited)
+            {
+                break;
+            }
```

Comment:
Is this actually needed? This wasn't there before, but was added when I was trying to externalize the control plane into a child process.

Reply:
yes?   if the main server is stopped/exited... we don't want to spin up the delay server at all, right?

Reply:
Well, the server sends a SIGTERM signal to the delay server. That causes the delay server to set a flag which instructs it to exit the processing loop.

Reply:
If the SIGTERM signal is sent when the server state transitions to stopped or exited and the flag is set, then this is basically equivalent to checking the flag, right? If that's the case, then it seems none of this diff hunk would be needed.

Reply:
That's true, but I think there's a case for grandpa crashing after the delay server is up. This code won't handle that and the SIGTERM-controlled flag won't either.

Maybe that isn't something we need to worry about?

If we do care about that, then we can add a check for the PPID being 1 or grandpa. If PPID is not who we expect, then shutdown. This solution would need to be checked in docker since containers may not always run with an init-like system.

Reply:
Of course, if grandpa crashes then maybe it's on the admin to kill the delay server and report it to us because that should never happen under normal circumstances.

Reply:
Oh, I didn't know that's what we were talking about.

I don't think we need to worry about that, personally... But if we want to be super safe I would think jotting down the PPID at the beginning and terminating if it changes would be the most appropriate action. Then the new main server process would just stand up a new delay server when this one goes away.

But like you said, this shouldn't happen. So, if it does, the admin should be treading very carefully...

Reply:
Captured in an issue: https://github.com/irods/irods/issues/6345

I'm going to remove those lines because the signal handler covers this.

Reply:
done